### PR TITLE
Add vsphere specific infrastructure config options

### PR DESCRIPTION
**Pull request by derekwaynecarr**
In order to protect the surface area of the API exposed via cloud.conf, we are going to pursue an alternate approach to mask the set of options we enable users to configure in their cloud.conf options that map to concepts we explicitly test.

This API enables integration of VSphere storage, which we will have to use to write a cloud.conf that is consumed by kube-apiserver, kube-controller-manager when platform=vsphere.

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
/cc @abhinavdahiya @gnufied @mfojtik @enxebre -- please take a look.

we need to agree where we can render this API into a generated cloud.conf.

File: config/v1/types_infrastructure.go [hunk 1]
```diff
@@ -22,12 +23,87 @@ type Infrastructure struct {
...
 type InfrastructureSpec struct {
-	// cloudConfig is a reference to a ConfigMap containing the cloud provider configuration file.
-	// This configuration file is used to configure the Kubernetes cloud provider integration
-	// when using the built-in cloud provider integration or the external cloud controller manager.
-	// The namespace for this config map is openshift-config.
+	// VSphereInfrastructureConfig specifies configuration when the cluster is installed on a VSphere platform.
+	// This is only settable if the platform type is VSphere.
 	// +optional
-	CloudConfig ConfigMapFileReference `json:"cloudConfig"`
+	VSphereInfrastructureConfig *VSphereInfrastructureConfig `json:"vsphereInfrastructureConfig,omitempty"`
```
Comment (abhinavdahiya) [inline config/v1/types_infrastructure.go]: nit: I would like to see somethings like:
```
spec:
   platform:
     vsphere:
        <here>
```

File: config/v1/types_infrastructure.go [hunk 2]
```diff
@@ -22,12 +23,87 @@ type Infrastructure struct {
...
+}
+
+// VSphereInfrastructureConfig specifies the configuration of one or more Virtual Centers where
+// the control plane and worker nodes are running in the cluster.  This information is used to
+// control the configuration of the vsphere cloud provider support in Kubernetes as supported
+// by OpenShift.
+type VSphereInfrastructureConfig struct {
+	// secretRef is name of the authentication secret for access to the VSphere infrastructure.
+	SecretRef corev1.SecretReference `json:"secretRef"`
```
Comment (abhinavdahiya) [inline config/v1/types_infrastructure.go]: nit: `SecretNameReference` so this is forced into a specified namespace.

File: config/v1/types_infrastructure.go [hunk 3]
```diff
@@ -22,12 +23,87 @@ type Infrastructure struct {
...
+
+	// port is the vCenter Server Port.
+	// defaults to 443 if not specified.
+	// +optional
+	Port *int32 `json:"port,omitempty"`
```
Comment (abhinavdahiya) [inline config/v1/types_infrastructure.go]: how difficult is to make users force `443`?

File: config/v1/types_infrastructure.go [hunk 4]
```diff
@@ -22,12 +23,87 @@ type Infrastructure struct {
...
+
+	// insecure if true means vCenter is configured with a self-signed cert.
+	// defaults to false.
+	// +optional
+	Insecure bool `json:"insecure,omitempty"`
```
Comment (abhinavdahiya) [inline config/v1/types_infrastructure.go]: how many users you do think have `insecure`. Would it be better to always require secure ?

**Comment by abhinavdahiya** [comment]
> we are going to pursue an alternate approach to mask the set of options we enable users to configure in their cloud.conf options that map to concepts we explicitly test.

do we have a reason to believe why we won't be able to support `cloud.conf` isn't that what k8s needs to always support?

**Comment by abhinavdahiya** [comment]
Also this putting entire vSphere cloud conf in infrastructure means entire API will need to change even when a platform specific field changes for cloud conf.

If we want this kind of detailed per platform spec it might be useful to look towards provider API like machines....

File: config/v1/types_infrastructure.go [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package v1
 
 import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+import corev1 "k8s.io/api/core/v1"
```
Comment (soltysh) [inline config/v1/types_infrastructure.go]: Please fix those imports.

File: config/v1/types_infrastructure.go (hunk 4 above)
Comment (gnufied) [reply config/v1/types_infrastructure.go]: A lot of them run with `insecure` default unfortunately.. :(

**Comment by gnufied** [comment]
Change itself looks like but there are 2 questions I have:

1. Will it become a chore to update this Type if upstream configuration changes?
2. Since we don't run e2e against vsphere, will we even notice the breakage if something changes?

File: config/v1/types_infrastructure.go (hunk 3 above)
Comment (derekwaynecarr) [reply config/v1/types_infrastructure.go]: i do not think we can force how a customer has configured their vsphere infrastructure.

File: config/v1/types_infrastructure.go (hunk 4 above)
Comment (derekwaynecarr) [reply config/v1/types_infrastructure.go]: again, i do not think we can require a change to infrastructure.

**Comment by derekwaynecarr** [comment]
we have requested upstream that the cloud.conf file formats be treated as a formal API.  this is being taken up by SIG Cloud.  the concern i have with cloud.conf being exposed directly is its a dumping ground for edge feature enablement that could break our cluster upgrades.  for a sampling, look at what is here:

gce cloud.conf
https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/gce/gce.go#L160

aws cloud.conf
https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L527

File: config/v1/types_infrastructure.go [hunk 6]
```diff
@@ -22,12 +23,87 @@ type Infrastructure struct {
...
+}
+
+// VSphereDiskConfig specifies configuration for
+type VSphereDiskConfig struct {
+	// scsiControllerType defines SCSI controller to be used.
+	SCSIControllerType string `json:"scsiControllerType"`
+}
+
+// VSphereWorkspaceConfig describes the endpoint used to create volumes.
+type VSphereWorkspaceConfig struct {
```
Comment (gnufied) [inline config/v1/types_infrastructure.go]: We need this field to create default storageclass and ensure a vsphere cluster comes with some kind of default storage.

**Comment by derekwaynecarr** [comment]
after further discussion, we are going to not use this pr and stick with the original api proposed.

**Pull request closed without merging**